Make Skia use bmalloc/fastMalloc

aperezdc opened the pull request:
#### 8722ec73f50ceaf96717b811b489bf3819db90a9
<pre>
Make Skia use bmalloc/fastMalloc
<a href="https://bugs.webkit.org/show_bug.cgi?id=269572">https://bugs.webkit.org/show_bug.cgi?id=269572</a>

Reviewed by Carlos Garcia Campos.

Implement functions used by Skia to handle memory allocations in terms
of WTF::fastMalloc() and friends.

* Source/ThirdParty/skia/CMakeLists.txt: Build the fastMalloc allocator
instead of the system-wide one when USE_SYSTEM_MALLOC is disabled.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp:
(WebCore::ImageBufferSkiaUnacceleratedBackend::create): Remove FIXME
comment, allocations done by Skia now go through bmalloc/fastMalloc.
* Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp: Added.
(sk_abort_no_print):
(sk_out_of_memory):
(sk_free):
(sk_realloc_throw):
(sk_malloc_flags):
(sk_malloc_size):
* Tools/Scripts/webkitpy/style/checker.py: Ignore lints for include
headers and underscores in identifiers for SkiaAllocatorFastMalloc.cpp
* Tools/ImageDiff/Skia.cmake: Add logic to avoid linking ImageDiff to
WTF when Skia has the bmalloc/fastMalloc allocator support built.

Canonical link: <a href="https://commits.webkit.org/275315@main">https://commits.webkit.org/275315@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8411cb33e0ce38a1f0a545b37048889b251ba4e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41520 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44088 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37612 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/43827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23641 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17864 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42094 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/17462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35761 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/41390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36765 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45457 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37086 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40852 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16335 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13412 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39258 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17954 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18010 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5546 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17598 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp:
```diff
@@ -0,0 +1,66 @@
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
```
Why is this LGPL?

aperezdc replied:
We use the LGPL in the API layer of the GTK and WPE ports and I have my editor automatically insert that as a template, which I forgot to change for this new source file.

Indeed, this is cross-port code so it should use the BSD license; I will change it.

carlosgcampos reviewed (approved):
Great, I think we can remove a couple of fixmes I added in places where we use skia to allocate pixels and we could use wrapper APIs to use fastmalloc and are no longer needed.

aperezdc committed:
Make Skia use bmalloc/fastMalloc
https://bugs.webkit.org/show_bug.cgi?id=269572

Reviewed by Carlos Garcia Campos.

Implement functions used by Skia to handle memory allocations in terms
of WTF::fastMalloc() and friends.

* Source/ThirdParty/skia/CMakeLists.txt: Build the fastMalloc allocator
instead of the system-wide one when USE_SYSTEM_MALLOC is disabled.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp:
(WebCore::ImageBufferSkiaUnacceleratedBackend::create): Remove FIXME
comment, allocations done by Skia now go through bmalloc/fastMalloc.
* Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp: Added.
(sk_abort_no_print):
(sk_out_of_memory):
(sk_free):
(sk_realloc_throw):
(sk_malloc_flags):
(sk_malloc_size):
* Tools/Scripts/webkitpy/style/checker.py: Ignore lints for include
headers and underscores in identifiers for SkiaAllocatorFastMalloc.cpp
* Tools/ImageDiff/Skia.cmake: Add logic to avoid linking ImageDiff to
WTF when Skia has the bmalloc/fastMalloc allocator support built.

Canonical link: https://commits.webkit.org/275315@main
